### PR TITLE
sgrep: init at 1.94a

### PR DESCRIPTION
**Pull request by eigengrau**
###### Description of changes

[sgrep](http://www.cs.helsinki.fi/~jjaakkol/sgrep.html) is a grep for structured text formats such as XML.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/text/sgrep/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchurl, runCommand, lib, m4, makeWrapper }:
+let
+  sgrep = stdenv.mkDerivation rec {
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:3]: `let` isn't needed here, just start with `stdenv.mkDerivation`.

File: pkgs/tools/text/sgrep/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      url =
+        "https://www.cs.helsinki.fi/pub/Software/Local/Sgrep/sgrep-${version}.tar.gz";
+      sha256 = "sha256-1bFkeOOrRHNeJCg9LYldLJyAE5yVIo3zvbKsRGOV+vk=";
+    };
+
+    buildInputs = [ makeWrapper ];
+
+    preFixup = ''
+      wrapProgram $out/bin/sgrep \
+        --prefix PATH : ${m4}/bin
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:17]:
```suggestion
        --prefix PATH : ${lib.makeBinPath [ m4 ]}
```

File: pkgs/tools/text/sgrep/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      expr='"<foo>" __ "</foo>"'
+      data="<foo>1</foo><bar>2</bar>"
+      ${sgrep}/bin/sgrep "$expr" <<<$data >$out
+      read result <$out
+      [[ $result = 1 ]]
+    '';
+
+    meta = with lib; {
+      homepage = "https://www.cs.helsinki.fi/u/jjaakkol/sgrep.html";
+      description = "a grep for structured text formats such as XML";
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:30]:
```suggestion
      description = "A grep for structured text formats such as XML";
```

File: pkgs/tools/text/sgrep/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchurl, runCommand, lib, m4, makeWrapper }:
+let
+  sgrep = stdenv.mkDerivation rec {
+    pname = "sgrep";
+    version = "1.94a";
+
+    src = fetchurl {
+      url =
+        "https://www.cs.helsinki.fi/pub/Software/Local/Sgrep/sgrep-${version}.tar.gz";
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:9]:
```suggestion
      url = "https://www.cs.helsinki.fi/pub/Software/Local/Sgrep/sgrep-${version}.tar.gz";
```

File: pkgs/tools/text/sgrep/default.nix [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      sha256 = "sha256-1bFkeOOrRHNeJCg9LYldLJyAE5yVIo3zvbKsRGOV+vk=";
+    };
+
+    buildInputs = [ makeWrapper ];
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:13]:
```suggestion
    nativeBuildInputs = [ makeWrapper ];
```

`makeWrapper` should be in `nativeBuildInputs`

File: pkgs/tools/text/sgrep/default.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      longDescription = ''
+        sgrep (structured grep) is a tool for searching and indexing text,
+        SGML, XML and HTML files and filtering text streams using
+        structural criteria.'';
+      platforms = platforms.unix;
+      license = licenses.gpl2Plus;
+      maintainers = with maintainers; [ eigengrau ];
+    };
+  };
+in sgrep
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:40]: See first comment.

**Review by moni-dz — CHANGES_REQUESTED** [review]

File: pkgs/tools/text/sgrep/default.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      [[ $result = 1 ]]
+    '';
+
+    meta = with lib; {
+      homepage = "https://www.cs.helsinki.fi/u/jjaakkol/sgrep.html";
+      description = "a grep for structured text formats such as XML";
+      longDescription = ''
+        sgrep (structured grep) is a tool for searching and indexing text,
+        SGML, XML and HTML files and filtering text streams using
+        structural criteria.'';
```
Comment (fortuneteller2k) [inline pkgs/tools/text/sgrep/default.nix:34]:
```suggestion
        structural criteria.
      '';
```

**Review by moni-dz — CHANGES_REQUESTED** [review]

File: pkgs/tools/text/sgrep/default.nix (hunk 7 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix]: Applied.

File: pkgs/tools/text/sgrep/default.nix (hunk 5 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix]: Yikes, applied.

File: pkgs/tools/text/sgrep/default.nix (hunk 4 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix]: Applied (editor was configured with nixfmt, not nixpkgs-fmt).

File: pkgs/tools/text/sgrep/default.nix (hunk 3 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix]: Applied.

File: pkgs/tools/text/sgrep/default.nix (hunk 1 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix:3]: We need a reference to the derivation output for `passthru.tests.smokeTest`. I figure one alternative would be:

```nix
passthru.tests.smokeTest = callPackage ({ sgrep }: runCommand "test-sgrep" { } ''
…
'') { };
```

… but this looked slightly clumsier than a let. Any recommendrations?

Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix:3]: Never mind, we can just add sgrep to the top-level callPackage arguments…

**Review by moni-dz — APPROVED** [review]

File: pkgs/tools/text/sgrep/default.nix [hunk 8]
```diff
@@ -0,0 +1,38 @@
+{ stdenv, sgrep, fetchurl, runCommand, lib, m4, makeWrapper }:
```
Comment (SuperSandro2000) [inline pkgs/tools/text/sgrep/default.nix:1]: Missing new line

File: pkgs/tools/text/sgrep/default.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+  src = fetchurl {
+    url = "https://www.cs.helsinki.fi/pub/Software/Local/Sgrep/sgrep-${version}.tar.gz";
+    sha256 = "sha256-1bFkeOOrRHNeJCg9LYldLJyAE5yVIo3zvbKsRGOV+vk=";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  preFixup = ''
+    wrapProgram $out/bin/sgrep \
```
Comment (SuperSandro2000) [inline pkgs/tools/text/sgrep/default.nix:14]: This should be done in postInstall. Also why do we need m4? That's really unusual.

Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix:14]: Applied.

> Also why do we need m4? That's really unusual.
sgrep shells out to m4 to pre-process expressions when the user wants to define and use custom macros.

File: pkgs/tools/text/sgrep/default.nix (hunk 8 above)
Comment (eigengrau) [reply pkgs/tools/text/sgrep/default.nix:1]: Ah, you mean a blank line between the arguments and the function body? Sure, let’s add it. Is this missing from the “coding conventions” section of the docs btw., or am I just overlooking it? If it’s missing I could try to add it later.

**Pull request merged**